In [1]:
import pickle
import pandas as pd
import numpy as np

import regex as re

from pybliometrics.scopus import AuthorSearch

from collections import Counter


### Step 0: Getting back to our groove

Before API testing will get familiar with the project & variables.

1- retr_q1

2- date_df_ok

3 - SumCetCet date_df_dicts

In [151]:
# 1-

%store -r retr_q1
retr_q1 = retr_q1

In [11]:
retr_q1.head()

Sbj_Area  \
Issn                 Year                                                      
(00011541, 15475905) 2010  {'Biochemistry, Genetics and Molecular Biology...   
                     2011  {'Chemical Engineering': 1, 'Environmental Sci...   
                     2012  {'Chemical Engineering': 1, 'Environmental Sci...   
                     2013  {'Biochemistry, Genetics and Molecular Biology...   
                     2014  {'Biochemistry, Genetics and Molecular Biology...   

                          CR_Results  SH_Results  CR_Count  Total_Docs  
Issn                 Year                                               
(00011541, 15475905) 2010       True        True       351         294  
                     2011       True       False       361         315  
                     2012       True       False       259         347  
                     2013       True        True       343         422  
                     2014       True        True       376         359

In [12]:
# 2-

%store -r date_df_ok
date_df_ok = date_df_ok

In [15]:
print(date_df_ok.shape)
date_df_ok.head()


(114759, 17)


reference-count publisher               published-print  \
DOI                                                                          
10.1002/aic.12400               26     Wiley   {'date-parts': [[2011, 7]]}   
10.1002/aic.12671               14     Wiley   {'date-parts': [[2012, 5]]}   
10.1002/aic.13810               37     Wiley   {'date-parts': [[2013, 1]]}   
10.1002/aic.14056                0     Wiley   {'date-parts': [[2013, 4]]}   
10.1002/aic.14601               37     Wiley  {'date-parts': [[2014, 11]]}   

                   is-referenced-by-count  \
DOI                                         
10.1002/aic.12400                      14   
10.1002/aic.12671                       0   
10.1002/aic.13810                      17   
10.1002/aic.14056                      19   
10.1002/aic.14601                      12   

                                                               title  \
DOI                                                                    
10.1002/aic.12400  [Fabrication of an anionic polythiophene/layer...   
10.1002/aic.12671  [Gas phase reaction kinetics in boron fibre pr...   
10.1002/aic.13810  [Rapid assembly of polyelectrolyte multilayer ...   
10.1002/aic.14056  [A computer-aided methodology to design safe f...   
10.1002/aic.14601  [Cake properties of nanocolloid evaluated by v...   

                                                              author  \
DOI                                                                    
10.1002/aic.12400  [{'given': 'Dongpeng', 'family': 'Yan', 'seque...   
10.1002/aic.12671  [{'given': 'Fatih', 'family': 'Fırat', 'sequen...   
10.1002/aic.13810  [{'given': 'Haiqi', 'family': 'Tang', 'sequenc...   
10.1002/aic.14056  [{'given': 'Phuong-Mai', 'family': 'Nguyen', '...   
10.1002/aic.14601  [{'given': 'Eiji', 'family': 'Iritani', 'seque...   

                                  published-online  container-title  \
DOI                                                                   
10.1002/aic.12400  {'date-parts': [[2010, 9, 17]]}  [AIChE Journal]   
10.1002/aic.12671   {'date-parts': [[2011, 6, 1]]}  [AIChE Journal]   
10.1002/aic.13810  {'date-parts': [[2012, 4, 23]]}  [AIChE Journal]   
10.1002/aic.14056   {'date-parts': [[2013, 3, 6]]}  [AIChE Journal]   
10.1002/aic.14601  {'date-parts': [[2014, 9, 16]]}  [AIChE Journal]   

                                            issued         ISSN  \
DOI                                                               
10.1002/aic.12400  {'date-parts': [[2010, 9, 17]]}  [0001-1541]   
10.1002/aic.12671   {'date-parts': [[2011, 6, 1]]}  [0001-1541]   
10.1002/aic.13810  {'date-parts': [[2012, 4, 23]]}  [0001-1541]   
10.1002/aic.14056   {'date-parts': [[2013, 3, 6]]}  [0001-1541]   
10.1002/aic.14601  {'date-parts': [[2014, 9, 16]]}  [0001-1541]   

                                                             subject  \
DOI                                                                    
10.1002/aic.12400  [General Chemical Engineering, Environmental E...   
10.1002/aic.12671  [General Chemical Engineering, Environmental E...   
10.1002/aic.13810  [General Chemical Engineering, Environmental E...   
10.1002/aic.14056  [General Chemical Engineering, Environmental E...   
10.1002/aic.14601  [General Chemical Engineering, Environmental E...   

                                         published  \
DOI                                                  
10.1002/aic.12400  {'date-parts': [[2010, 9, 17]]}   
10.1002/aic.12671   {'date-parts': [[2011, 6, 1]]}   
10.1002/aic.13810  {'date-parts': [[2012, 4, 23]]}   
10.1002/aic.14056   {'date-parts': [[2013, 3, 6]]}   
10.1002/aic.14601  {'date-parts': [[2014, 9, 16]]}   

                                                             Results  \
DOI                                                                    
10.1002/aic.12400  [https://sci.bban.top/pdf/10.1002/aic.12400.pd...   
10.1002/aic.12671  [https://sci.bban.top/pdf/10.1002/aic.12671.pd...   
10.100

In [17]:
# 3-

with open("cet_date_df_dict", "rb") as fp:
    cet_date_df_ok = pickle.load(fp)

with open("sum_date_df_dict", "rb") as fp:
    sum_date_df_ok = pickle.load(fp)

with open("ogi_date_df_dict", "rb") as fp:
    ogi_date_df_ok = pickle.load(fp)

cet_date_df = pd.DataFrame(cet_date_df_ok)
sum_date_df = pd.DataFrame(sum_date_df_ok)
ogi_date_df = pd.DataFrame(ogi_date_df_ok)

date_df_5b = pd.concat([cet_date_df,sum_date_df,ogi_date_df])

with open("date_df_5b","wb") as fp:
    pickle.dump(date_df_5b, fp)

In [23]:
date_df_5b

DOI  \
0            10.1002/aic.14056   
1            10.1002/aic.14601   
2            10.1002/aic.14847   
3            10.1002/aic.15322   
4            10.1002/aic.16019   
...                        ...   
12509  10.3945/ajcn.2010.29422   
12510  10.3945/ajcn.111.022343   
12511      10.1093/ajcn/nqy217   
12512      10.1093/ajcn/nqz173   
12513     10.1093/ajcn/nqaa281   

                                              direct_url  \
0      https://sci.bban.top/pdf/10.1002/aic.14056.pdf...   
1      https://sci.bban.top/pdf/10.1002/aic.14601.pdf...   
2      https://sci.bban.top/pdf/10.1002/aic.14847.pdf...   
3      https://backend.orbit.dtu.dk/ws/files/12385525...   
4      https://sci.bban.top/pdf/10.1002/aic.16019.pdf...   
...                                                  ...   
12509  https://sci.bban.top/pdf/10.3945/ajcn.2010.294...   
12510  https://sci.bban.top/pdf/10.3945/ajcn.111.0223...   
12511  https://sci.bban.top/pdf/10.1093/ajcn%252Fnqy2...   
12512  https://sci.bban.top/pdf/10.1093/ajcn%252Fnqz1...   
12513  https://sci.bban.top/pdf/10.1093/ajcn%252Fnqaa...   

                               Date           Affiliation  \
0              [July 3, 2012, None]                France   
1             [Mar. 27, 2014, None]                 Japan   
2      [Feb 12, 2015, Apr 17, 2015]  United Arab Emirates   
3      [Apr 02, 2016, May 16, 2016]               Denmark   
4               [May 1, 2017, None]         United States   
...                             ...                   ...   
12509     [February 23, 2010, None]        United Kingdom   
12510         [June 23, 2011, None]           Netherlands   
12511          [May 17, 2018, None]           Netherlands   
12512     [February 26, 2019, None]                Greece   
12513     [February 26, 2020, None]              Malaysia   

                                                Keywords  
0      [(foodpackaging, 0.384), (contamination, 0.377...  
1      [(porosity, 0.3551), (pressures, 0.3424), (cak...  
2      [(nanofluids, 0.4791), (nanofluid, 0.4535), (c...  
3      [(publication, 0.4136), (process, 0.3538), (co...  
4      [(optimization, 0.4064), (knapsackproblem, 0.3...  
...                                                  ...  
12509  [(diabetes, 0.3765), (triglyceride, 0.3656), (...  
12510  [(hypertension, 0.4852), (nutrition, 0.4168), ...  
12511  [(apolipoproteins, 0.436), (apolipoprotein, 0....  
12512  [(dietary, 0.4073), (cancers, 0.3767), (carcin...  
12513  [(dietary, 0.4995), (diet, 0.4386), (nutrition...  

[36514 rows x 5 columns]

It is seen that date_df_5b was run only for already successful articles & so that the results can be compared & saved accordingly. Before moving on to the API part, will merge the 2 dfs & as well as simplifying & updating date_df_ok for future use & avoid confusion


In [ ]:

# def result_mapper(lst):
#     if len(lst) == 2:
#         return lst[0], np.nan, np.nan
#     elif len(lst) == 3:
#         return lst[0], lst[1], lst[2]

# date_df_ok["direct_url"], date_df_ok["Rec_date"], date_df_ok["Acc_date"] = zip(*date_df_ok.Results.map(result_mapper))

In [29]:
#  Getting ready date_df_5b for merging:

date_df_5b.set_index("DOI",inplace=True, drop=True)
date_df_5b.drop("direct_url", axis=1, inplace=True)

date_df_5b.rename(columns={"Date": "New_dates"}, inplace=True)

In [58]:
date_test = date_df_5b.merge(date_df_ok["Rec_date"],how="left",left_index=True, right_index=True)

date_test.loc[date_test.Rec_date.notna(),"Rec_date"] = date_test[date_test.Rec_date.notna()].Rec_date.map(lambda x: re.sub(' +', ' ', x))

In [101]:
len(date_test[(date_test.Rec_date.str.strip() != date_test.Date.str[0].str.strip()) & (date_test.Rec_date.notna()) & (date_test.Date != "cant_read_pdf") & (date_test.Date != "direct_url_error") ][["Date","Rec_date"]])

183

In [100]:
print(f'Why different?\nNew date found: {date_test.loc["10.1016/j.trsl.2009.06.009","Date"][0]}\nRec_date found: {date_test.loc["10.1016/j.trsl.2009.06.009","Rec_date"]}')

Why different?
New date found: May 17, 2009
Rec_date found: May17, 2009


It is seen that out of 36K rows, there are 183 rows with different received dates found. However, upon further inspection as below, it is seen that the actual number is much lower, as most "differences" are caused by whitespaces or tabs in the columns. Therefore can write a simple function to map these two columns accordingly & depending on errors & earlier dates, etc. However, this can de skipped for now, as it is not certain that all the article records here will also be used in q1_artcl_df. There are more crucial tasks at hand for now!

In [106]:
# Getting date_df_ok ready:

# Getting rid of duplicated columns 
date_df_ok.drop(["first_author","direct_url","Rec_date","Acc_date"], axis=1, inplace=True)

# Merging the datasets:
date_df_ok_final = date_df_ok.merge(date_df_5b, how="left", left_index=True, right_index=True)

# Re-ordering columns:
date_df_ok_final = date_df_ok_final[["ISSN", "issued", "container-title", "publisher", "title", "author", "subject", "reference-count", "is-referenced-by-count", "published", "published-print", "published-online", "Results", "New_dates", "Keywords","Affiliation"]]

In [117]:
with open("date_df_ok_final","wb") as fp:
    pickle.dump(date_df_ok_final, fp)

Now, everything is simplified & ready for next steps. We have 2 main variables:

* date_df_ok_final -> that has everything related to the 144K articles that were retrieved so far
* retr_q1 -> everything related to Q1 journals

### Part 2: Scopus API & Function

In [9]:
with open("date_df_ok_final", "rb") as fp:
    date_df_ok_final = pickle.load(fp)

In [50]:
print(f"Subject NaNs: {len(date_df_ok_final[date_df_ok_final.subject.isna()])}")

Subject NaNs: 326


In [71]:
from itertools import chain

sbj_list = list()
date_df_ok_final.subject.map(lambda x: sbj_list.append(x) if type(x) == list else False)

def flatten(listOfLists):
    "Flatten one level of nesting"
    return chain.from_iterable(listOfLists)

sbj_set = set(flatten(sbj_list))


len(sbj_set)

346

Fromt the test above it is seen that the "subject" column from CR is actually not useful as it is not standardized to Scopus sbj. areas & there are 346 unique values as opposed to 27. This shows that:

a- This field wont be useful during analysis & Scimago sbj. areas from jrnl_df should be used intead -> another reason for a proper jrnl & artcl mapping!

b- Cant use this column during Scopus Author Search API query, and need to find another method

For now, will get all author results from the API & can use jrnl_df sbj. areas later to filter the results if there are multiple author returns!

---

* drop subject col. from article dataset!
* map sbj_area from jrnl_df to artcl_df using (ISSN & issued.year)
* create a dict from elsevier API guide


In [ ]:
# Adding an ID column to retr_q1:

retr_q1["ID"] = pd.factorize(retr_q1.index)[0]

In [129]:
def Scopus_author(author_list):
    for aut in author_list:
        name = aut["given"]
        surname = aut["family"]
        s = AuthorSearch(f'AUTHLAST({surname}) and AUTHFIRST({name})')
        if s.get_results_size()>0:
            aut["country"] = s.authors[0]._asdict()["country"]
        else:
            aut["country"] = None
    return author_list

In [ ]:
# After adding the country field for every author in "author" column:
# Can use to create a seperate "author_countries" column

def country_counter(author_col):
    return Counter([aut["country"] for aut in author_col])


However, during testing it is seen that we need to add sbj. areas to the author search query as there are too many results right now. 

In [130]:
author_test = date_df_ok_final[["author","subject"]][:10]

author_test["test"] = author_test.author.map(Scopus_author)

author_test["country_test"] = author_test.test.map(country_counter)

In [143]:
author_test["country_test"]

DOI
10.1002/aic.12400                {'China': 5, 'United Kingdom': 1}
10.1002/aic.12671                                    {'Turkey': 4}
10.1002/aic.13810                                     {'China': 3}
10.1002/aic.14056                {'United States': 1, 'France': 3}
10.1002/aic.14601                        {'Japan': 3, 'Taiwan': 2}
10.1002/aic.14847                      {'United Arab Emirates': 1}
10.1002/aic.15322    {'Denmark': 2, 'South Korea': 1, 'Mexico': 1}
10.1002/aic.16019                             {'United States': 2}
10.1002/aic.16379               {'Australia': 1, 'Switzerland': 1}
10.1002/aic.16792           {'United States': 1, 'South Korea': 3}
Name: country_test, dtype: object

In [144]:
s_s = AuthorSearch('AUTHLAST(Evans) and AUTHFIRST(David G.) AND (SUBJAREA(MEDI) OR SUBJAREA(BIOC) OR SUBJAREA(NEUR))')

In [103]:
s_f = AuthorSearch('AUTHLAST(kokes) and AUTHFIRST(oguz)')

In [145]:
s_s.get_results_size()

15

In [187]:
# Creating a "new" ISSN column for date_df_ok that is compatible with retr_q1 ISSNs:

date_df_ok_final["ISSN_new"] = date_df_ok_final.ISSN.map(lambda x: [iss.replace("-","").strip() for iss in x]).map(tuple)

Using the ISSN_new & issued columns, we can map journals in retr_q1 to articles in date_df with journal IDs

This is crucial as this will simplify many processes in the future, including mapping Sbj. Areass of journals to articles that will be used during Author Search API!

In [227]:
def one_filter(row, jrnl_df):
    id_list = jrnl_df[jrnl_df.index.map(lambda x: (row.ISSN_new[0] in x[0]) & (x[1] == row.issued["date-parts"][0][0]))]["ID"].tolist()
    return id_list


In [229]:
def two_filter(row, jrnl_df):
    id_list = jrnl_df[jrnl_df.index.map(lambda x: (row.ISSN_new[0] in x[0]) & (row.ISSN_new[1] in x[0]) & (x[1] == row.issued["date-parts"][0][0]))]["ID"].tolist()
    return id_list


In [231]:
def three_filter(row, jrnl_df):
    id_list = jrnl_df[jrnl_df.index.map(lambda x: (row.ISSN_new[0] in x[0]) & (row.ISSN_new[1] in x[0]) & (row.ISSN_new[2] in x[0]) & (x[1] == row.issued["date-parts"][0][0]))]["ID"].tolist()
    return id_list


In [ ]:
date_df_ok_final.loc[date_df_ok_final.ISSN_new.map(len)==1,"jrnl_id"] = date_df_ok_final[date_df_ok_final.ISSN_new.map(len)==1].apply(lambda x: one_filter(x, retr_q1), axis=1)

In [ ]:
date_df_ok_final.loc[date_df_ok_final.ISSN_new.map(len)==2,"jrnl_id"] = date_df_ok_final[date_df_ok_final.ISSN_new.map(len)==3].apply(lambda x: two_filter(x, retr_q1), axis=1)

In [ ]:
date_df_ok_final.loc[date_df_ok_final.ISSN_new.map(len)==3,"jrnl_id"] = date_df_ok_final[date_df_ok_final.ISSN_new.map(len)==3].apply(lambda x: three_filter(x, retr_q1), axis=1)

After 200 mins, the one_filter run is cancelled, which was for around 34K rows. Instead the sbj_area_query creator can be tested individually using the scopus_codes_dict below.

This also affects the potential use for date_df_ok_final usage in q1_artcl_df as we need a certain mapping between articles & journals to ensure data integrity & easier analysis futhter down the line. This will be review again at a later time.

In [238]:
scopus_codes_dict = {'Agricultural and Biological Sciences': 'AGRI',
 'Arts and Humanities': 'ARTS',
 'Biochemistry, Genetics and Molecular Biology': 'BIOC',
 'Business, Management and Accounting': 'BUSI',
 'Chemical Engineering': 'CENG',
 'Chemistry': 'CHEM',
 'Computer Science': 'COMP',
 'Decision Sciences': 'DECI',
 'Dentistry': 'DENT',
 'Earth and Planetary Sciences': 'EART',
 'Economics, Econometrics and Finance': 'ECON',
 'Energy': 'ENER',
 'Engineering': 'ENGI',
 'Environmental Science': 'ENVI',
 'Health Professions': 'HEAL',
 'Immunology and Microbiology': 'IMMU',
 'Materials Science': 'MATE',
 'Mathematics': 'MATH',
 'Medicine': 'MEDI',
 'Neuroscience': 'NEUR',
 'Nursing': 'NURS',
 'Pharmacology, Toxicology and Pharmaceutics': 'PHAR',
 'Physics and Astronomy': 'PHYS',
 'Psychology': 'PSYC',
 'Social Sciences': 'SOCI',
 'Veterinary': 'VETE',
 'Multidisciplinary': 'MULT'}

In [236]:
def sbj_area_query_creator(sbj_list):
    query = " AND ".join([f"SUBJAREA({scopus_codes_dict[sbj]})" for sbj in sbj_list])
    return query

In [239]:
# Testing:

sbj_area_query_creator(["Veterinary", "Psychology"])

'SUBJAREA(VETE) AND SUBJAREA(PSYC)'

In [262]:
# Testing 2:

#retr_q1.loc[retr_q1.index.map(lambda x: ("17531055" in x[0]) & (x[1] == 2018)),"Sbj_Area"][0].keys()
sbj_area_query_creator(retr_q1.loc[retr_q1.index.map(lambda x: ("17531055" in x[0]) & (x[1] == 2018)),"Sbj_Area"][0].keys())

'SUBJAREA(SOCI) AND SUBJAREA(ARTS)'